llo/plugin.go

```diff
@@ -0,0 +1,924 @@
+			0,
+			nil,
+			nil,
+			nil,
+		}
+		return json.Marshal(outcome)
+	}
+
+	/////////////////////////////////
+	// Decode previousOutcome
+	/////////////////////////////////
+	var previousOutcome Outcome
+	if err := json.Unmarshal(outctx.PreviousOutcome, &previousOutcome); err != nil {
+		return nil, fmt.Errorf("error unmarshalling previous outcome: %v", err)
+	}
+
+	/////////////////////////////////
+	// Decode observations
+	/////////////////////////////////
+
+	// a single valid retirement report is enough
+	var validPredecessorRetirementReport *RetirementReport
+
+	shouldRetireVotes := 0
+
+	timestampsNanoseconds := []int64{}
+
+	removeChannelVotesByID := map[commontypes.ChannelID]int{}
+
+	// for each channelId count number of votes that mention it and count number of votes that include it.
+	addChannelVotesByHash := map[ChannelHash]int{}
+	addChannelDefinitionsByHash := map[ChannelHash]ChannelDefinitionWithID{}
+
+	streamObservations := map[commontypes.StreamID][]*big.Int{}
+
+	for _, ao := range aos {
+		observation := Observation{}
+		// TODO: Use protobufs
+		if err := json.Unmarshal(ao.Observation, &observation); err != nil {
+			p.Logger.Warnw("ignoring invalid observation", "oracleID", ao.Observer, "error", err)
+			continue
+		}
+
+		if len(observation.AttestedPredecessorRetirement) != 0 && validPredecessorRetirementReport == nil {
+			pcd := *p.PredecessorConfigDigest
+			retirementReport, err := p.PredecessorRetirementReportCache.CheckAttestedRetirementReport(pcd, observation.AttestedPredecessorRetirement)
+			if err != nil {
+				p.Logger.Warnw("ignoring observation with invalid attested predecessor retirement", "oracleID", ao.Observer, "error", err, "predecessorConfigDigest", pcd)
+				continue
+			}
+			validPredecessorRetirementReport = &retirementReport
+		}
+
+		if observation.ShouldRetire {
+			shouldRetireVotes++
+		}
+
+		timestampsNanoseconds = append(timestampsNanoseconds, observation.UnixTimestampNanoseconds)
+
+		for channelID := range observation.RemoveChannelIDs {
+			removeChannelVotesByID[channelID]++
+		}
+
+		for channelID, channelDefinition := range observation.AddChannelDefinitions {
+			defWithID := ChannelDefinitionWithID{channelDefinition, channelID}
+			channelHash := MakeChannelHash(defWithID)
+			addChannelVotesByHash[channelHash]++
+			addChannelDefinitionsByHash[channelHash] = defWithID
+		}
+
+		for id, obsResult := range observation.StreamValues {
+			if obsResult.Valid {
+				streamObservations[id] = append(streamObservations[id], obsResult.Val)
+			} else {
+				p.Logger.Debugw("Ignoring invalid observation", "streamID", id, "oracleID", ao.Observer)
+			}
+		}
+	}
+
+	if len(timestampsNanoseconds) == 0 {
+		return nil, errors.New("no valid observations")
+	}
+
+	var outcome Outcome
+
+	/////////////////////////////////
+	// outcome.LifeCycleStage
+	/////////////////////////////////
+	if previousOutcome.LifeCycleStage == LifeCycleStageStaging && validPredecessorRetirementReport != nil {
```

Comment:
what if we don't want to promote to LifeCylceStageProduction? is there a way to cancel - is that just canceling the job and creating new configs on-chain?

Reply:
The LifeCycleStage code currently does nothing since it is not even implemented on the contract. It was copied direct from lorenz's sketch and I didn't touch any of it. It will be implemented in https://smartcontract-it.atlassian.net/browse/MERC-3386